### PR TITLE
Add Swin2SR

### LINKED ISSUE #19568: Add Swin2SR
### Model description

Swin2SR is a Swinv2-based model for image super resolution and compression.

### Open source status

- [X] The model implementation is available
- [X] The model weights are available

### Provide useful links for the implementation

https://github.com/mv-lab/swin2sr

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

Fixes #19568 and replaces #19667 

This PR adds Swin2SR, a Swinv2-based model for image super resolution, compression and restoration.

To do:

- [x] finish `Swin2SRImageProcessor` - should incorporate padding
- [x] fix integration test
- [x] transfer checkpoints to the appropriate organization

### CODE REVIEW

File: src/transformers/__init__.py [hunk 1]
```diff
@@ -2049,17 +2070,6 @@
             "load_tf_weights_in_t5",
         ]
     )
-    _import_structure["models.switch_transformers"].extend(
```
Comment (NielsRogge) [inline src/transformers/__init__.py:2052]: I put this in the appropriate alphabetical place

File: docs/source/en/_toctree.yml [hunk 2]
```diff
@@ -44,7 +44,8 @@
       title: Use tokenizers from 🤗 Tokenizers
     - local: multilingual
       title: Inference for multilingual models
-    - sections:
```
Comment (NielsRogge) [inline docs/source/en/_toctree.yml:47]: It's very weird that this gets updated, I copied from main and only added swin2sr, but this gets deleted.

I guess it happens when running `make fix-copies`

**Comment by NielsRogge** [comment]
Gently pinging @sgugger here

File: docs/source/en/_toctree.yml (hunk 2 above)
Comment (sgugger) [reply docs/source/en/_toctree.yml:47]: You should revert this part.

File: docs/source/en/model_doc/swin2sr.mdx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+In this paper, we explore the novel Swin Transformer V2, to improve SwinIR for image super-resolution, and in particular, the compressed input scenario. Using this method we can tackle the major issues in training transformer vision models, such as training instability, resolution gaps between pre-training and fine-tuning, and hunger on data. We conduct experiments on three representative tasks: JPEG compression artifacts removal, image super-resolution (classical and lightweight), and compressed image super-resolution. Experimental results demonstrate that our method, Swin2SR, can improve the training convergence and performance of SwinIR, and is a top-5 solution at the "AIM 2022 Challenge on Super-Resolution of Compressed Image and Video".*
+
+Tips:
+
+<INSERT TIPS ABOUT MODEL HERE>
```
Comment (sgugger) [inline docs/source/en/model_doc/swin2sr.mdx:28]: Needs to be filled.

File: src/transformers/__init__.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/__init__.py:2052]: It already is. Not sure what branch you are on but a rebase/merge seems necessary.

File: src/transformers/models/swin2sr/modeling_swin2sr.py [hunk 4]
```diff
@@ -0,0 +1,1155 @@
...
+_FEAT_EXTRACTOR_FOR_DOC = "AutoFeatureExtractor"
+
+# Base docstring
+_CHECKPOINT_FOR_DOC = "caidas/swin2sr-classicalsr-x2-64"
+_EXPECTED_OUTPUT_SHAPE = [1, 64, 768]
+
+
+SWIN2SR_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    # TODO update to appropriate organization
+    "nielsr/swin2SR-classical-sr-x2-64",
```
Comment (sgugger) [inline src/transformers/models/swin2sr/modeling_swin2sr.py:55]: The config used caidas here.

File: src/transformers/models/swin2sr/modeling_swin2sr.py [hunk 5]
```diff
@@ -0,0 +1,1155 @@
...
+
+        return BaseModelOutput(
+            last_hidden_state=sequence_output,
+            hidden_states=encoder_outputs.hidden_states,
+            attentions=encoder_outputs.attentions,
+        )
+
+
+# TODO inherit from nn.Module here
+class Upsample(nn.Sequential):
```
Comment (sgugger) [inline src/transformers/models/swin2sr/modeling_swin2sr.py:960]: Still to do.

File: src/transformers/models/swin2sr/modeling_swin2sr.py [hunk 6]
```diff
@@ -0,0 +1,1155 @@
...
+        elif scale == 3:
+            m.append(nn.Conv2d(num_features, 9 * num_features, 3, 1, 1))
+            m.append(nn.PixelShuffle(3))
+        else:
+            raise ValueError(f"scale {scale} is not supported. Supported scales: 2^n and 3.")
+        super(Upsample, self).__init__(*m)
+
+
+# TODO inherit from nn.Module here
+class UpsampleOneStep(nn.Sequential):
```
Comment (sgugger) [inline src/transformers/models/swin2sr/modeling_swin2sr.py:983]: Same here.

**Review by sgugger — COMMENTED** [review]
Thanks for adding this model!

File: src/transformers/models/swin2sr/image_processing_swin2sr.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+        pad_height = (old_height // size + 1) * size - old_height
+        pad_width = (old_width // size + 1) * size - old_width
+        image = np.concatenate([image, np.flip(image, [0])], 0)[: old_height + pad_height, :, :]
+        image = np.concatenate([image, np.flip(image, [1])], 1)[:, : old_width + pad_width, :]
+
+        if data_format is not None:
+            image = to_channel_dimension_format(image, data_format)
+
+        return image
```
Comment (NielsRogge) [inline src/transformers/models/swin2sr/image_processing_swin2sr.py:113]: cc @amyeroberts I've implemented padding as follows to match the original implementation, let me know if that's ok.

i'm not using pad from image_transforms atm

Comment (amyeroberts) [inline src/transformers/models/swin2sr/image_processing_swin2sr.py:113]: I've just checked, and we can get equivalent outputs with the image transforms pad:

```
import requests

import numpy as np
from PIL import Image

from transformers.image_transforms import pad, get_image_size

url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)
image = np.array(image)

size = 8
old_height, old_width = get_image_size(image)
pad_height = (old_height // size + 1) * size - old_height
pad_width = (old_width // size + 1) * size - old_width

flip_image = image.copy()
flip_image = np.concatenate([flip_image, np.flip(flip_image, [0])], 0)[: old_height + pad_height, :, :]
flip_image = np.concatenate([flip_image, np.flip(flip_image, [1])], 1)[:, : old_width + pad_width, :]

pad_image = pad(image, ((0, pad_height), (0, pad_width)), mode="symmetric")

assert (flip_image == pad_image).all()
```

Note: the transforms in the library are written to handle the different channel dimension formats so you don't need to change the input image to e.g. get the (height, width) dimensions. 

```suggestion
        old_height, old_width = get_image_size(image)
        pad_height = (old_height // size + 1) * size - old_height
        pad_width = (old_width // size + 1) * size - old_width
        image = pad(image, ((0, pad_height), (0, pad_width)), mode="symmetric", data_format=data_format)
        return image
```

Comment (NielsRogge) [reply src/transformers/models/swin2sr/image_processing_swin2sr.py]: Thanks I've updated the code!

**Review by sgugger — APPROVED** [review]
